src/Http/Headers/src/HeaderUtilities.cs

```diff
@@ -625,34 +625,34 @@ public static StringSegment UnescapeAsQuotedString(StringSegment input)
         {
             input = RemoveQuotes(input);
 
-            // First pass to calculate the size of the InplaceStringBuilder
+            // First pass to calculate the size of the string
             var backSlashCount = CountBackslashesForDecodingQuotedString(input);
 
             if (backSlashCount == 0)
             {
                 return input;
             }
 
-            var stringBuilder = new InplaceStringBuilder(input.Length - backSlashCount);
-
-            for (var i = 0; i < input.Length; i++)
+            return string.Create(input.Length - backSlashCount, input, (span, segment) =>
             {
-                if (i < input.Length - 1 && input[i] == '\\')
+                var spanIndex = 0;
+                for (var i = 0; i < segment.Length; i++)
                 {
-                    // If there is an backslash character as the last character in the string,
-                    // we will assume that it should be included literally in the unescaped string
-                    // Ex: "hello\\" => "hello\\"
-                    // Also, if a sender adds a quoted pair like '\\''n',
-                    // we will assume it is over escaping and just add a n to the string.
-                    // Ex: "he\\llo" => "hello"
-                    stringBuilder.Append(input[i + 1]);
-                    i++;
-                    continue;
+                    if (i < segment.Length - 1 && segment[i] == '\\')
```

Comment:
Add `&& (uint)spanIndex < (uint)span.Length` so the JIT may elide the bound checks for `span[spanIndex++]`? The loop will be compacter, than with the two emitted bound checks.

Reply:
Why do you have to cast to uint? var i = 0; i < segment.Length  works in ints.

Reply:
The JIT can't figure out tha bound checks can be eliminated, so these are hints for the JIT.

`(uint)spanIndex < (uint)spanLength` because the JIT can't know if indexing into the Span is safe. So we check it, then it's fine and we get only one "check" instead of two inside the loop body, thus making the loop tighter.

`(uint)j < (uint)segment.Length` here `j = i + 1` is checked explicitely, as the JIT won't / can't figure out wheter `i + 1` is within bounds or not.

Although still with these hints the genereted code isn't ideal. When I try `char[]` instead of `ROS<char>` the emitted code is good, and the span based version should be as close as possible.